src/librbd/operation/DisableFeaturesRequest.cc

```diff
@@ -580,12 +580,40 @@ Context *DisableFeaturesRequest<I>::handle_update_flags(int *result) {
   return nullptr;
 }
 
+template <typename I>
+void DisableFeaturesRequest<I>::send_unblock_writes() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  {
+    std::unique_lock locker{image_ctx.owner_lock};
+
+    if (image_ctx.exclusive_lock != nullptr && m_requests_blocked) {
+      image_ctx.exclusive_lock->unblock_requests();
+      m_requests_blocked = false;
+    }
+
+    if (m_writes_blocked) {
+      image_ctx.io_image_dispatcher->unblock_writes();
+      m_writes_blocked = false;
+    }
+  }
+
+  if (!m_prepare_lock_completed) {
+    image_ctx.state->handle_prepare_lock_complete();
+    m_prepare_lock_completed = true;
+  }
+}
+
 template <typename I>
 void DisableFeaturesRequest<I>::send_notify_update() {
   I &image_ctx = this->m_image_ctx;
   CephContext *cct = image_ctx.cct;
   ldout(cct, 20) << this << " " << __func__ << dendl;
 
+  send_unblock_writes();
```

Comment:
I think this is wrong.  Currently the order is:
1. prepare lock
2. block writes
3. block requests
4. acquire exclusive lock
5. [...] send_notify_update [...]
4. release exclusive lock
3. unblock requests
2. unblock writes
1. prepare lock complete

Note that "exit" order is exactly the reverse of the "enter" order.  With this change, everything is unblocked before the lock is released.

Reply:
I'm not quite sure I understand your point. But if you mean to call `release_exclusive_lock` before `send_notify_update` and call the `unblock requests/writes` followed by `prepare lock complete` at the end, then it will lead to canceling of the actual async request.

Here are some of my old logs:
```
2023-04-26T16:49:12.869+0530 7f61c97fa640 20 librbd::DisableFeaturesRequest: 0x7f61d0009420 handle_remove_journal: r=0                                                                                                                  
2023-04-26T16:49:12.869+0530 7f61c97fa640 20 librbd::DisableFeaturesRequest: 0x7f61d0009420 send_remove_object_map                                                                                                                      
2023-04-26T16:49:12.869+0530 7f61c97fa640 20 librbd::DisableFeaturesRequest: 0x7f61d0009420 send_set_features: new_features=61, features_mask=64                                                                                        
2023-04-26T16:49:12.869+0530 7f61c97fa640 20 librbd::DisableFeaturesRequest: 0x7f61d0009420 send_append_op_event                                                                                                                        
2023-04-26T16:49:12.873+0530 7f620148d640 20 librbd::DisableFeaturesRequest: 0x7f61d0009420 handle_set_features: r=0                                                                                                                    
2023-04-26T16:49:12.873+0530 7f620148d640 20 librbd::DisableFeaturesRequest: 0x7f61d0009420 send_release_exclusive_lock                                                                                                                 
2023-04-26T16:49:12.873+0530 7f620148d640 10 librbd::ManagedLock: 0x7f61d0000b78 release_lock:                                                                                                                                          
2023-04-26T16:49:12.873+0530 7f620148d640 10 librbd::ManagedLock: 0x7f61d0000b78 send_release_lock:                                                                                                                                     
2023-04-26T16:49:12.873+0530 7f620148d640 10 librbd::ExclusiveLock: 0x7f61d0000b60 pre_release_lock_handler                                                                                                                             
2023-04-26T16:49:12.873+0530 7f620148d640 10 librbd::exclusive_lock::PreReleaseRequest: 0x7f61d40131e0 send_cancel_op_requests:                                                                                                         
2023-04-26T16:49:12.873+0530 7f620148d640 10 librbd::ImageCtx: canceling async requests: count=1                                                                                                                                        
2023-04-26T16:49:12.873+0530 7f620148d640 10 librbd::ImageCtx: canceling async request: 0x7f61d0009420                                                                                                                                  
2023-04-26T16:49:42.844+0530 7f6200c8c640  5 librbd::Watcher: 0x7f61d4007840 notifications_blocked: blocked=0                                                                                                                           
2023-04-26T16:49:42.844+0530 7f6200c8c640 10 librbd::Watcher::C_NotifyAck 0x7f61f0002060 C_NotifyAck: id=206158430213, handle=140058145419040                                                                                           
2023-04-26T16:49:42.844+0530 7f6200c8c640 10 librbd::ImageWatcher: 0x7f61d4007840 remote update_features request: [4306,140418385724656,1] 64 disabled                                                                                  
2023-04-26T16:49:42.844+0530 7f6200c8c640 20 librbd::ImageWatcher: handle_operation_request                                                                                                                                             
2023-04-26T16:49:42.844+0530 7f6200c8c640 20 librbd::ImageWatcher: handle_operation_request PrasannaXXX: has m_image_ctx.exclusive_lock                                                                                                 
2023-04-26T16:49:42.844+0530 7f6200c8c640 20 librbd::ExclusiveLock: 0x7f61d0000b60 accept_request=0 (request_type=0)                                                                                                                    
2023-04-26T16:49:42.844+0530 7f6200c8c640 20 librbd::ImageWatcher: handle_operation_request PrasannaXXX: doesn't has m_image_ctx.exclusive_lock                                                                                         
2023-04-26T16:49:42.844+0530 7f6200c8c640 10 librbd::Watcher::C_NotifyAck 0x7f61f0002060 finish: r=0                                                                                                                                    
2023-04-26T16:49:42.904+0530 7f620148d640  5 librbd::Watcher: 0x7f61d4007840 notifications_blocked: blocked=0                                                                                                                           
2023-04-26T16:49:42.904+0530 7f620148d640 10 librbd::Watcher::C_NotifyAck 0x7f61b00011f0 C_NotifyAck: id=206158430214, handle=140058145419040                                                                                           
2023-04-26T16:49:42.904+0530 7f620148d640 10 librbd::ImageWatcher: 0x7f61d4007840 remote update_features request: [4306,140418385724656,1] 64 disabled                                                                                  
2023-04-26T16:49:42.904+0530 7f620148d640 20 librbd::ImageWatcher: handle_operation_request                                                                                                                                             
2023-04-26T16:49:42.904+0530 7f620148d640 20 librbd::ImageWatcher: handle_operation_request PrasannaXXX: has m_image_ctx.exclusive_lock                                                                                                 
2023-04-26T16:49:42.904+0530 7f620148d640 20 librbd::ExclusiveLock: 0x7f61d0000b60 accept_request=0 (request_type=0)                                                                                                                    
2023-04-26T16:49:42.904+0530 7f620148d640 20 librbd::ImageWatcher: handle_operation_request PrasannaXXX: doesn't has m_image_ctx.exclusive_lock                                                                                         
2023-04-26T16:49:42.904+0530 7f620148d640 10 librbd::Watcher::C_NotifyAck 0x7f61b00011f0 finish: r=0                                                                                                                                    
2023-04-26T16:49:42.911+0530 7f6200c8c640  5 librbd::Watcher: 0x7f61d4007840 notifications_blocked: blocked=0                                                                                                                           
2023-04-26T16:49:42.911+0530 7f6200c8c640 10 librbd::Watcher::C_NotifyAck 0x7f61f0009660 C_NotifyAck: id=206158430215, handle=140058145419040                                                                                           

```

Reply:
@idryomov could you please revisit this?

Reply:
@idryomov can you take a look on this please?

Reply:
I think the idea that we discussed in a meeting was to call `send_notify_update` at the end of `DisableFeaturesRequest`, after releasing exclusive lock, unblocking requests, unblocking writes and calling `handle_prepare_lock_complete`.  Basically move it from step 5 to a new, its own, step 10.

>     1. prepare lock 
>     2. block writes
>     3. block requests
>     4. acquire exclusive lock
>     5. [...] send_notify_update [...]
>     6. release exclusive lock
>     7. unblock requests
>     8. unblock writes
>     9. prepare lock complete

Reply:
What I noticed before is that calling release_exclusive_lock before send_notify_update will lead to canceling of the actual async request. Its been quite sometime, let me work on it tomorrow and get more thoughts.

Reply:
@idryomov Could you please comment here on why you think this version is not acceptable (the reason is out if my head now, sorry), also for the history sake. I have a new version which is based on @trociny comments before. I will push it as soon as you update here. Thanks!

Reply:
> @idryomov Could you please comment here on why you think this version is not acceptable (the reason is out of my head now, sorry), also for the history sake.

Well, we block writes and stop accepting lock requests before acquiring the lock.  This is needed to avoid someone stealing the lock from under our nose.  It seems only logical to unwind in reverse order: after releasing the lock, start accepting lock requests and unblock writes.  It's also what the existing code does.

I didn't say that what you proposed was unacceptable -- just that the ordering became weird.   But the details are out of my head too -- this is a problem with splitting discussion between PR comments and meetings ;)

> I have a new version which is based on @trociny comments before.

Addressing this in rbd-nbd seems nicer anyway!  The remote request processing code is pretty convoluted and bending librbd to provide an execution environment for completely arbitrary callbacks is unlikely to be worth the effort.

Reply:
+1 For fixing this in rbd-nbd. I seemed to write about this earlier but can't find the comment now. We may just declare that it is unsafe to call any librbd function from inside `librbd::UpdateWatchCtx::handle_notify()` callback, as it may got stuck. So what `NBDWatchCtx::handle_notify()` should do is just a set a flag that an update has happened and wakeup a thread, that will do the real job. The same way as we do with quiesce notifications (`NBDQuiesceWatchCtx`).

Reply:
@trociny I was referring to the below comment

> I am not sure if we want to go so far though, i.e. should we care so much for notification to be answered by all potential watchers, which might be misbehaving. May be we just want to send the header update notification and proceed. In this case we may want to return to @pkalever's initial approach, i.e. don't call notification flush (but instead of modifying the `flush` operation making it NOP, we could just remove `m_image_ctx.state->flush_update_watchers` in `ImageWatcher<I>::handle_payload(const HeaderUpdatePayload)`.


This works as far as this issue is concerned,

```
[0] pkalever 😎 build✨ git diff ../src/librbd/ImageWatcher.cc
diff --git a/src/librbd/ImageWatcher.cc b/src/librbd/ImageWatcher.cc
index f3405102fda..3124530ac2d 100644
--- a/src/librbd/ImageWatcher.cc
+++ b/src/librbd/ImageWatcher.cc
@@ -970,12 +970,8 @@ bool ImageWatcher<I>::handle_payload(const HeaderUpdatePayload &payload,
                                     C_NotifyAck *ack_ctx) {
   ldout(m_image_ctx.cct, 10) << this << " image header updated" << dendl;
 
   m_image_ctx.state->handle_update_notification();
   m_image_ctx.perfcounter->inc(l_librbd_notify);
-  if (ack_ctx != nullptr) {
-    m_image_ctx.state->flush_update_watchers(new C_ResponseMessage(ack_ctx));
-    return false;
-  }
   return true;
 }
``` 

unless there are any side effects which I don't see yet.

Reply:
Removing `flush_update_watchers` from `handle_payload` may be a good idea in general (at least it looks good to me unless it breaks something), but even if it fixes this particular issue, I still think it is good idea to have a restriction for `librbd::UpdateWatchCtx::handle_notify()` callbacks not to call any librbd functions for this image context. Because it will not work in general case (consider a case when you decide to call `write` or `flush` from the callback). So we would need to provide a list of functions (like `stat`) which are safe to call, and add tests that would check that they still work after any code change. Or we can just declare that no librbd function can be called from inside the notify context. 

Actually I think I will be fine with any approach, though I would prefer the last one.

Reply:
@pkalever I interpreted your "I have a new version which is based on @trociny comments before" to mean https://github.com/ceph/ceph/pull/50593#discussion_r1170323366 because that was the alternative discussed in a meeting.  My apologies for any confusion here!

> Alternative approach could be just to document that update notification callbacks are not allowed to call any librbd code to avoid potential deadlock and then modify rbd-nbd to just set a flag in the callback and process it in a separate thread as we already do with quiesce callbacks.

Reply:
No problem @idryomov 

Thank you @trociny & @idryomov. If you think we should be looking at a direction of fixing this at rbd-nbd level, I'm all ok. I will look into it and propose new changes.